File: src/UnwindMC.Tests/AstTests.cs
using NDis86;
using NUnit.Framework;
using System.Collections.Generic;
using UnwindMC.Analysis.Ast;
using UnwindMC.Analysis.Data;
using UnwindMC.Analysis.IL;
using static UnwindMC.Tests.Helpers.AstHelper;
using static UnwindMC.Tests.Helpers.FlowHelper;
using static UnwindMC.Tests.Helpers.ILHelper;

namespace UnwindMC.Tests
{
    [TestFixture]
    public class AstTests
    {
        [Test]
        public void TestAstWithFunctionPointers()
        {
            var asn0 = Assign(Register(OperandType.ESI), Stack(0));
            var cmp0 = Compare(Register(OperandType.ESI), Stack(4));
            var asn1 = Assign(Register(OperandType.EAX), Pointer(OperandType.ESI));
            var cmp1 = Compare(Register(OperandType.EAX), Value(0));
            var call = Call(Register(OperandType.EAX));
            var add = Add(Register(OperandType.ESI), Value(4));
            var ret = Return();

            asn0.SetVariableIds(0, -1);
            cmp0.SetVariableIds(0, -1);
            asn1.SetVariableIds(1, 0);
            cmp1.SetVariableIds(1, -1);
            call.SetVariableIds(1, -1);
            add.SetVariableIds(0, -1);
            ret.SetVariableIds(-1, -1);

            asn0.AddDefaultChild(cmp0);
            cmp0.AddDefaultChild(ret);
            cmp0.AddConditionalChild(ILBranchType.Less, asn1);
            asn1.AddDefaultChild(cmp1);
            cmp1.AddDefaultChild(add);
            cmp1.AddConditionalChild(ILBranchType.NotEqual, call);
            call.AddDefaultChild(add);
            add.AddDefaultChild(cmp0);

            var blocks = Blocks(
                Sequential(asn0),
                While(cmp0,
                    Sequential(asn1),
                    Conditional(cmp1,
                        Blocks(Sequential(call)),
                        Blocks(Sequential())),
                    Sequential(add)),
                Sequential(ret));

            var parameterTypes = new List<Type>
            {
                new Type(true, 1),
                new Type(true, 1),
            };

            var variableTypes = new List<Type>
            {
                new Type(true, 1),
                new Type(true, 0),
            };

            var ast = new AstBuilder(blocks, parameterTypes, variableTypes).BuildAst();
            var expected = Scope(
                Assign(Var("var0"), Var("arg0")),
                While(Less(Var("var0"), Var("arg1")),
                    Scope(
                        Assign(Var("var1"), Dereference(Var("var0"))),
                        IfThenElse(NotEqual(Var("var1"), Val(0)),
                            Scope(Call(Var("var1"))),
                            Scope()),
                        Assign(Var("var0"), Add(Var("var0"), Val(1))))),
                Ret());
            AssertAstEqual(expected, ast);
        }

        [Test]
        public void TestAstFindMax()
        {
            var asn0 = Assign(Register(OperandType.ECX), Stack(4));
            var asn1 = Assign(Register(OperandType.EAX), Value(int.MinValue));
            var cmp0 = Compare(Register(OperandType.ECX), Value(0));
            var asn2 = Assign(Register(OperandType.EDX), Stack(0));
            var asn3 = Assign(Register(OperandType.EAX), Value(int.MinValue));
            var asn4 = Assign(Register(OperandType.ESI), Pointer(OperandType.EDX));
            var cmp1 = Compare(Register(OperandType.EAX), Register(OperandType.ESI));
            var asn5 = Assign(Register(OperandType.EAX), Register(OperandType.ESI));
            var add0 = Add(Register(OperandType.EDX), Value(4));
            var sub0 = Subtract(Register(OperandType.ECX), Value(1));
            var cmp2 = Compare(Register(OperandType.ECX), Value(0));
            var ret = Return();

            asn0.SetVariableIds(1, -1);
            asn1.SetVariableIds(0, -1);
            cmp0.SetVariableIds(1, -1);
            asn2.SetVariableIds(2, -1);
            asn3.SetVariableIds(0, -1);
            asn4.SetVariableIds(3, 2);
            cmp1.SetVariableIds(0, 3);
            asn5.SetVariableIds(0, 3);
            add0.SetVariableIds(2, -1);
            sub0.SetVariableIds(1, -1);
            cmp2.SetVariableIds(1, -1);
            ret.SetVariableIds(-1, 0);

            asn0.AddDefaultChild(asn1);
            asn1.AddDefaultChild(cmp0);
            cmp0.AddDefaultChild(ret);
            cmp0.AddConditionalChild(ILBranchType.NotEqual, asn2);
            asn2.AddDefaultChild(asn3);
            asn3.AddDefaultChild(asn4);
            asn4.AddDefaultChild(cmp1);
            cmp1.AddDefaultChild(add0);
            cmp1.AddConditionalChild(ILBranchType.Less, asn5);
            asn5.AddDefaultChild(add0);
            add0.AddDefaultChild(sub0);
            sub0.AddDefaultChild(cmp2);
            cmp2.AddDefaultChild(ret);
            cmp2.AddConditionalChild(ILBranchType.NotEqual, asn4);

            var blocks = Blocks(
                Sequential(
                    asn0,
                    asn1),
                Conditional(cmp0,
                    Blocks(
                        Sequential(
                            asn2,
                            asn3),
                        DoWhile(cmp2,
                            Sequential(asn4),
                            Conditional(cmp1,
                                Blocks(Sequential(asn5)),
                                Blocks(Sequential())),
                            Sequential(
                                add0,
                                sub0))),
                    Blocks(Sequential())),
                Sequential(ret));

            var parameterTypes = new List<Type>
            {
                new Type(false, 1),
                new Type(false, 0),
            };

            var variableTypes = new List<Type>
            {
                new Type(false, 0),
                new Type(false, 0),
                new Type(false, 1),
                new Type(false, 0),
            };

            var ast = new AstBuilder(blocks, parameterTypes, variableTypes).BuildAst();
            var expected = Scope(
                Assign(Var("var0"), Var("arg1")),
                Assign(Var("var1"), Val(int.MinValue)),
                IfThenElse(NotEqual(Var("var0"), Val(0)),
                    Scope(
                        Assign(Var("var2"), Var("arg0")),
                        Assign(Var("var1"), Val(int.MinValue)),
                        DoWhile(
                            Scope(
                                Assign(Var("var3"), Dereference(Var("var2"))),
                                IfThenElse(Less(Var("var1"), Var("var3")),
                                    Scope(Assign(Var("var1"), Var("var3"))),
                                    Scope()),
                                Assign(Var("var2"), Add(Var("var2"), Val(1))),
                                Assign(Var("var0"), Subtract(Var("var0"), Val(1)))),
                            NotEqual(Var("var0"), Val(0)))),
                    Scope()),
                Ret(Var("var1")));
            AssertAstEqual(expected, ast);
        }
    }
}

File: src/UnwindMC.Tests/AstTests.fs
[<NUnit.Framework.TestFixture>]
module AstTests

open System
open NDis86
open NUnit.Framework
open Ast
open FlowAnalyzer
open IL
open Type

[<Test>]
let testAstWithFunctionPointers (): unit =
    let nop0 = Nop
    let asn0 = Assign { left = Register OperandType.ESI; right = Stack 0; leftId = 0; rightId = -1 }
    let cmp0 = Compare { left = Register OperandType.ESI; right = Stack 4; leftId = 0; rightId = -1 }
    let br0 = Branch <| branch GreaterOrEqual 10uL
    let asn1 = Assign { left = Register OperandType.EAX; right = Pointer (OperandType.ESI, 0); leftId = 1; rightId = 0 }
    let cmp1 = Compare { left = Register OperandType.EAX; right = Value 0; leftId = 1; rightId = -1 }
    let br1 = Branch <| branch Equal 8uL
    let call = Call { operand = Register OperandType.EAX; operandId = 1 }
    let add = Add { left = Register OperandType.ESI; right = Value 4; leftId = 0; rightId = -1 }
    let br3 = Branch <| branch Unconditional 2uL
    let nop1 = Nop
    let ret = Return { operand = Register OperandType.EAX; operandId = -1 }

    let blocks =
        [|
            SequentialBlock { instructions = [| nop0; asn0 |] }
            WhileBlock {
                condition = invertCondition [| cmp0; br0 |]
                body =
                    [|
                        SequentialBlock { instructions = [| asn1 |] }
                        ConditionalBlock {
                            condition = invertCondition [| cmp1; br1 |]
                            trueBranch = [| SequentialBlock { instructions = [| call |] } |]
                            falseBranch = [||]
                        }
                        SequentialBlock { instructions = [| add |] }
                    |]
            }
            SequentialBlock { instructions = [| nop1; ret |] }
        |]

    let parameterTypes =
        [|
            { isFunction = true; indirectionLevel = 1; size = 4 }
            { isFunction = true; indirectionLevel = 1; size = 4 }
        |]

    let variableTypes =
        [|
            { isFunction = true; indirectionLevel = 1; size = 4 }
            { isFunction = true; indirectionLevel = 0; size = 4 }
        |]

    let func = AstBuilder.buildAst "" blocks parameterTypes [||] variableTypes
    let expected =
        [|
            Assignment (Var "var0", VarRef (Var "arg0"))
            While
                (
                    Binary (Operator.Less, VarRef (Var "var0"), VarRef (Var "arg1")),
                    [|
                        Assignment (Var "var1", Dereference (VarRef (Var "var0")))
                        IfThenElse
                            (
                                Binary (Operator.NotEqual, VarRef (Var "var1"), Expression.Value 0),
                                [| FunctionCall (VarRef (Var "var1")) |],
                                [||]
                            )
                        Assignment (Var "var0", Binary (Operator.Add, VarRef (Var "var0"), Expression.Value 1))
                    |]
                )
            Statement.Return None
        |]
    AstHelper.assertAstEqual expected func.body

[<Test>]
let testAstFindMax (): unit =
    let nop0 = Nop
    let asn0 = Assign { left = Register OperandType.ECX; right = Stack 4; leftId = 1; rightId = -1 }
    let asn1 = Assign { left = Register OperandType.EAX; right = Value Int32.MinValue; leftId = 0; rightId = -1 }
    let cmp0 = Compare { left = Register OperandType.ECX; right = Value 0; leftId = 1; rightId = -1 }
    let br0 = Branch <| branch Equal 15uL
    let asn2 = Assign { left = Register OperandType.EDX; right = Stack 0; leftId = 2; rightId = -1 }
    let asn3 = Assign { left = Register OperandType.EAX; right = Value Int32.MinValue; leftId = 0; rightId = -1 }
    let asn4 = Assign { left = Register OperandType.ESI; right = Pointer (OperandType.EDX, 0); leftId = 3; rightId = 2 }
    let cmp1 = Compare { left = Register OperandType.EAX; right = Register OperandType.ESI; leftId = 0; rightId = 3 }
    let br1 = Branch <| branch GreaterOrEqual 11uL
    let asn5 = Assign { left = Register OperandType.EAX; right = Register OperandType.ESI; leftId = 0; rightId = 3 }
    let add0 = Add { left = Register OperandType.EDX; right = Value 4; leftId = 2; rightId = -1 }
    let sub0 = Subtract { left = Register OperandType.ECX; right = Value 1; leftId = 1; rightId = -1 }
    let cmp2 = Compare { left = Register OperandType.ECX; right = Value 0; leftId = 1; rightId = -1 }
    let br2 = Branch <| branch NotEqual 7uL
    let nop1 = Nop
    let ret = Return { operand = Register OperandType.EAX; operandId = 0 }

    let blocks =
        [|
            SequentialBlock { instructions = [| nop0; asn0; asn1 |] }
            ConditionalBlock {
                condition = invertCondition [| cmp0; br0 |]
                trueBranch =
                    [|
                        SequentialBlock { instructions = [| asn2; asn3 |] }
                        DoWhileBlock {
                            condition = [| cmp2; br2 |]
                            body =
                                [|
                                    SequentialBlock { instructions = [| asn4 |] }
                                    ConditionalBlock {
                                        condition = invertCondition [| cmp1; br1 |]
                                        trueBranch = [| SequentialBlock { instructions = [| asn5 |] } |]
                                        falseBranch = [||]
                                    }
                                    SequentialBlock { instructions = [| add0; sub0 |] }
                                |]
                        }
                    |]
                falseBranch = [||]
            }
            SequentialBlock { instructions = [| nop1; ret |] }
        |]

    let parameterTypes =
        [|
            { isFunction = false; indirectionLevel = 1; size = 4 }
            { isFunction = false; indirectionLevel = 0; size = 4 }
        |]

    let variableTypes =
        [|
            { isFunction = false; indirectionLevel = 0; size = 4 }
            { isFunction = false; indirectionLevel = 0; size = 4 }
            { isFunction = false; indirectionLevel = 1; size = 4 }
            { isFunction = false; indirectionLevel = 0; size = 4 }
        |]

    let func = AstBuilder.buildAst "" blocks parameterTypes [||] variableTypes
    let expected =
        [|
            Assignment (Var "var0", VarRef (Var "arg1"))
            Assignment (Var "var1", Expression.Value Int32.MinValue)
            IfThenElse
                (
                    Binary (Operator.NotEqual, VarRef (Var "var0"), Expression.Value 0),
                    [|
                        Assignment (Var "var2", VarRef (Var "arg0"))
                        Assignment (Var "var1", Expression.Value Int32.MinValue)
                        DoWhile
                            (
                                [|
                                    Assignment (Var "var3", Dereference (VarRef (Var "var2")))
                                    IfThenElse
                                        (
                                            Binary (Operator.Less, VarRef (Var "var1"), VarRef (Var "var3")),
                                            [| Assignment (Var "var1", VarRef (Var "var3")) |],
                                            [||]
                                        )
                                    Assignment (Var "var2", Binary (Operator.Add, VarRef (Var "var2"), Expression.Value 1))
                                    Assignment (Var "var0", Binary (Operator.Subtract, VarRef (Var "var0"), Expression.Value 1))
                                |],
                                Binary (Operator.NotEqual, VarRef (Var "var0"), Expression.Value 0)
                            )
                    |],
                    [||]
                )
            Statement.Return (Some (Var "var1"))
        |]
    AstHelper.assertAstEqual expected func.body
